### PR TITLE
Allow setting cluster Id explicitly

### PR DESCRIPTION
**Pull request by feedmeapples**
**What changed?**

Added a config option that allows to explicitly set Cluster ID through config

**Why?**

allow to temporal CLI that runs in in-memory mode to persist the ID between runs

**How did you test it?**

manually + in tests with https://github.com/temporalio/cli/pull/96

**Potential risks**

**Is hotfix candidate?**

no

### CODE REVIEW

File: common/cluster/metadata.go [hunk 1]
```diff
@@ -104,6 +104,8 @@ type (
 		InitialFailoverVersion int64 `yaml:"initialFailoverVersion"`
 		// Address indicate the remote service address(Host:Port). Host can be DNS name.
 		RPCAddress string `yaml:"rpcAddress"`
+		// Cluster ID allows to explicitly the ID of the cluster. Optional.
+		ClusterID  string `yaml:"clusterId"`
```
Comment (feedmeapples) [inline common/cluster/metadata.go:108]: maybe this should be ignored too if set in .yaml files?

Comment (bergundy) [reply common/cluster/metadata.go:108]: I think that'd be great, I don't see a reason for anyone to set this but our CLI.

File: temporal/fx.go [hunk 2]
```diff
@@ -662,11 +662,21 @@ func ApplyClusterMetadataConfigProvider(
...
+			// Only configure current cluster metadata from static config file
 			continue
 		}
 
-		// Only configure current cluster metadata from static config file
-		clusterId := uuid.New()
+		var clusterId string
+		if uuid.Parse(clusterInfo.ClusterID) == nil {
+			if clusterInfo.ClusterID != "" {
+				logger.Warn("Cluster ID in Cluster Metadata config is not a valid uuid. Generated new Cluster ID. ")
```
Comment (feedmeapples) [inline temporal/fx.go:673]:
```suggestion
				logger.Warn("Cluster Id in Cluster Metadata config is not a valid uuid. Generating a new Cluster Id. ")
```

File: temporal/fx.go [hunk 3]
```diff
@@ -662,11 +662,21 @@ func ApplyClusterMetadataConfigProvider(
...
+			// Only configure current cluster metadata from static config file
 			continue
 		}
 
-		// Only configure current cluster metadata from static config file
-		clusterId := uuid.New()
+		var clusterId string
+		if uuid.Parse(clusterInfo.ClusterID) == nil {
+			if clusterInfo.ClusterID != "" {
+				logger.Warn("Cluster Id in Cluster Metadata config is not a valid uuid. Generating a new Cluster Id. ")
```
Comment (feedmeapples) [inline temporal/fx.go:673]:
```suggestion
				logger.Warn("Cluster Id in Cluster Metadata config is not a valid uuid. Generating a new Cluster Id ")
```

File: temporal/fx.go [hunk 4]
```diff
@@ -662,11 +662,21 @@ func ApplyClusterMetadataConfigProvider(
...
+			// Only configure current cluster metadata from static config file
 			continue
 		}
 
-		// Only configure current cluster metadata from static config file
-		clusterId := uuid.New()
+		var clusterId string
+		if uuid.Parse(clusterInfo.ClusterID) == nil {
+			if clusterInfo.ClusterID != "" {
+				logger.Warn("Cluster Id in Cluster Metadata config is not a valid uuid. Generating a new Cluster Id ")
```
Comment (feedmeapples) [inline temporal/fx.go:673]:
```suggestion
				logger.Warn("Cluster Id in Cluster Metadata config is not a valid uuid. Generating a new Cluster Id")
```

File: common/cluster/metadata.go [hunk 5]
```diff
@@ -104,6 +104,8 @@ type (
 		InitialFailoverVersion int64 `yaml:"initialFailoverVersion"`
 		// Address indicate the remote service address(Host:Port). Host can be DNS name.
 		RPCAddress string `yaml:"rpcAddress"`
+		// Cluster ID allows to explicitly set the ID of the cluster. Optional.
+		ClusterID  string `yaml:"-"`
```
Comment (yux0) [inline common/cluster/metadata.go:108]: will it load it from config? if yes, use `yaml:"clusterId"`

Comment (feedmeapples) [reply common/cluster/metadata.go:108]: it's intentionally ignored https://github.com/temporalio/temporal/pull/3883#discussion_r1093896936

This is used by Temporal single binary

**Review by yux0 — APPROVED** [review]

**Pull request merged**